# **NOTEBOOK 13**
---

# **Modelos del lenguaje basados en redes neuronales artificiales**

## **Redes neuronales recurrentes (RNN)**

### **Ejemplo de aplicación de un modelo LSTM a la clasificación de textos**

#### **Dataset**

El conjunto de datos "AG_NEWS" es un conjunto de datos de clasificación de texto ampliamente utilizado en el campo del procesamiento de lenguaje natural (NLP). Contiene noticias de diferentes categorías y se utiliza comúnmente para tareas de clasificación de texto. El conjunto de datos AG_NEWS consta de noticias de cuatro categorías principales, que son:

1. **World**: Noticias sobre eventos y acontecimientos globales, como política internacional, relaciones internacionales y noticias mundiales en general.

2. **Sports**: Noticias relacionadas con eventos deportivos, resultados de partidos, eventos deportivos nacionales e internacionales, etc.

3. **Business**: Noticias relacionadas con el mundo de los negocios, finanzas, economía, empresas, informes de ganancias y otros temas económicos.

4. **Sci/Tech**: Noticias relacionadas con ciencia y tecnología, incluyendo avances científicos, novedades tecnológicas, gadgets, investigaciones científicas y más.

Cada instancia del conjunto de datos AG_NEWS generalmente consiste en un título y un cuerpo de una noticia, junto con una etiqueta que indica la categoría a la que pertenece. 

In [4]:
from torchtext import datasets
from torchtext.data import to_map_style_dataset
import numpy as np

# Load the dataset
train_iter, test_iter = datasets.AG_NEWS(split=('train', 'test'))

train_ds = to_map_style_dataset(train_iter)
test_ds = to_map_style_dataset(test_iter)

train = np.array(train_ds)
test = np.array(test_ds)

#### **Tokenización**

In [5]:
# Create vocabulary and embedding

from torchtext.vocab import build_vocab_from_iterator
from torchtext.data.utils import get_tokenizer

tokenizer = get_tokenizer("basic_english")

vocab = build_vocab_from_iterator(map(lambda x: tokenizer(x[1]), train_iter), specials=['<pad>','<unk>'])
vocab.set_default_index(vocab["<unk>"])

In [11]:
print("Tamaño del vocabulario:", len(vocab), "tokens")
print("Tokenización de la frase 'Here is an example sentence':", tokenizer("Here is an example sentence"))
print("Índices de las palabras 'here', 'is', 'an', 'example', 'supercalifragilisticexpialidocious':", vocab(['here', 'is', 'an', 'example', 'supercalifragilisticexpialidocious']))
print("Palabras correspondientes a los índices 475, 21, 30, 5297, 0:", vocab.lookup_tokens([475, 21, 30, 5297, 0]))
print("Las diez primeras palabras del vocabulario:", vocab.get_itos()[:10])

Tamaño del vocabulario: 95812 tokens
Tokenización de la frase 'Here is an example sentence': ['here', 'is', 'an', 'example', 'sentence']
Índices de las palabras 'here', 'is', 'an', 'example', 'supercalifragilisticexpialidocious': [476, 22, 31, 5298, 1]
Palabras correspondientes a los índices 475, 21, 30, 5297, 0: ['version', 'at', 'from', 'establish', '<pad>']
Las diez primeras palabras del vocabulario: ['<pad>', '<unk>', '.', 'the', ',', 'to', 'a', 'of', 'in', 'and']


In [12]:
text_pipeline = lambda x: vocab(tokenizer(x))
label_pipeline = lambda x: int(x) - 1

print("Tokenización de la frase 'Here is an example sentence':", text_pipeline("Here is an example sentence"))

Tokenización de la frase 'Here is an example sentence': [476, 22, 31, 5298, 2994]


#### **DataLoader**

En PyTorch, **DataLoader** se utiliza para cargar y manejar datos de manera eficiente durante el entrenamiento de modelos de aprendizaje profundo, especialmente en tareas de aprendizaje supervisado como la clasificación, la regresión y más. El DataLoader forma parte de la biblioteca torch.utils.data, y su objetivo principal es facilitar la administración de lotes (batches) de datos y la distribución de esos lotes al modelo de forma automática.

La función <code>collate_batch</code> es una función personalizada que se utiliza como argumento para collate_fn al crear instancias de los objetos DataLoader. Tiene la responsabilidad de procesar y agrupar las muestras individuales de datos dentro de un lote (batch) de manera que sean compatibles para su posterior procesamiento dentro de la LSTM. Es especialmente útil cuando las secuencias de texto tienen longitudes diferentes y es necesario realizar un relleno (padding) para que todas tengan la misma longitud.

In [13]:
from torch.utils.data import DataLoader
import torch


def collate_batch(batch):
    label_list, text_list = [], []
    for sample in batch:
        label, text = sample
        text_list.append(torch.tensor(text_pipeline(text), dtype=torch.long))
        label_list.append(label_pipeline(label))
    return torch.tensor(label_list, dtype=torch.long), torch.nn.utils.rnn.pad_sequence(text_list, batch_first=True, padding_value=vocab["<pad>"])

train_dataloader = DataLoader(
    train_iter, batch_size=64, shuffle=True, collate_fn=collate_batch
)

test_dataloader = DataLoader(
    test_iter, batch_size=64, shuffle=True, collate_fn=collate_batch
)

Para verificar que estamos creando correctamente los lotes, vamos a imprimir las primeras cuatro instancias:

In [19]:
for batch in train_dataloader:
    print(batch[1][:4])
    print("\n")
    print(batch[0][:4])
    print("\n")
    break

tensor([[   31,    59,  2345,    18,    94,    38,    31,   366, 23082,     8,
         23823,   346,     3, 29304,    19,     3,    59, 11043,     7,   354,
          1746,     4,     3,  1503,     7,     3,    59,  5318,    11,     3,
           348,    22,   588,     5,    38,   390,   712,  4173,    73,     8,
          1602,  5599,     2,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0],
        [20520, 24355,  1478,    94,    38,   885,   124,   216,   745,     8,
            23,   556,    64, 21336,     2,    18,    13,    10,  1807,     5,
             6,    24,   774,    18,    94, 12500, 21741,     7,     3, 15420,
         42558, 23487,     4,     6,  5214,     7, 24355,   720,     2,     0,
             0,     0,     0

#### **Modelo LSTM**

Definimos una clase llamada `LSTMTextClassificationModel` para la clasificación de texto basado en LSTM. Esta clase hereda de `nn.Module`, la clase base para todos los modelos en PyTorch.

**Inicialización del Modelo**: En el método `__init__`, se definen los componentes principales del modelo y se configuran sus parámetros:
   - **vocab_size**: El tamaño del vocabulario, es decir, la cantidad de palabras únicas en el conjunto de datos de entrenamiento.
   - **embed_dim**: La dimensión de los vectores de embedding para representar las palabras.
   - **hidden_dim**: El tamaño de la capa oculta de la LSTM, que controla la cantidad de unidades o "memoria" en la red LSTM.
   - **num_class**: El número de clases de salida en la tarea de clasificación.


**Capa de Embedding**: Se define una capa de embedding (`nn.Embedding`) que se utilizará para representar las palabras como vectores densos. Esta capa no es pre-entrenada, lo que significa que los vectores de embedding se entrenarán junto con el modelo durante el proceso de entrenamiento.

**Capa LSTM**: Se define una capa LSTM (`nn.LSTM`) que toma los vectores de embedding como entrada. `embed_dim` es la dimensión de entrada de la capa LSTM, y `hidden_dim` es la dimensión de su capa oculta. `batch_first=True` indica que los datos de entrada tendrán la forma `(batch_size, sequence_length, embed_dim)`.

**Capa Fully Connected (FC)**: Se define una capa completamente conectada (`nn.Linear`) que se utiliza para producir las salidas de clasificación. Toma las salidas de la capa LSTM correspondientes a la última iteración y las reduce a `num_class` dimensiones, que corresponde al número de clases en la tarea de clasificación.


In [20]:
import torch
import torch.nn as nn

class LSTMTextClassificationModel(nn.Module):
    def __init__(self, vocab_size, embed_dim, hidden_dim, num_class):
        super(LSTMTextClassificationModel, self).__init__()
        self.embedding = nn.Embedding(vocab_size, embed_dim)  # <-- Capa de embedding genérica (no pre-entrenada)
        self.lstm = nn.LSTM(embed_dim, hidden_dim, batch_first=True)
        self.fc = nn.Linear(hidden_dim, num_class)

    def forward(self, text):
        embedded = self.embedding(text)  # <-- Tras pasar por la capa de embedding, las palabras se representan como vectores
        lstm_out, _ = self.lstm(embedded)
        # Tomar la última salida de la secuencia LSTM
        last_output = lstm_out[:, -1, :]
        output = self.fc(last_output)
        return output

Antes de seguir veamos cómo se conforma un *batch* de datos tras pasar la capa de *embedding*. Como se muestra en la figura siguiente, un *batch* es un vector de tres dimensiones, donde la primera dimensión es el tamaño del *batch*, la segunda dimensión es el número de palabras en cada secuencia y la tercera dimensión es el tamaño del vector de *embedding* (*channels* o *features*).

![alt text](imgs/Lote1_b.svg)

Por ejemplo, un *token* estaría ahora representado por un vector de *embedding* de tres componentes, en lugar de un escalar. Si lo quisiéramos referenciar sería: <code>mini_lote[0,3,:]</code>.

![alt text](imgs/Lote2_b.svg)


Fíjate que la tercera dimensión del tensor de salida de la LSTM no tiene por qué tener el mismo tamaño que la tercera dimensión del tensor de entrada. La parte del tensor correspondiente a la línea de código <code>last_output = lstm_out[:, -1, :]</code> sería parecida a la siguiente:

![alt text](imgs/Lote3_b.svg)

Ya podemos pasar un primer mini-lote de datos por la red. Para ello, vamos a crearlo a partir de los datos de entrenamiento. Esto solo lo hacemos para asegurarnos de que el código funciona correctamente.

In [21]:
model = LSTMTextClassificationModel(len(vocab), 32, 64, 4)
model.train()

for batch in train_dataloader:
    predicted_label = model(batch[1])
    label = batch[0]
    break

print(batch[1][:4])
print(predicted_label[:4])
print(label[:4])

tensor([[  203,   938,   207, 12578,   982,   203,   938,   207, 12578, 76146,
           982,    29,    82,   921,    64,   203,   303,     7,     3, 23174,
         75444,    18,   215,   227,    12,     8,     3,  7751, 12578,  1566,
           471,     2,   113,    86,   203,    34,   368,    39,     5,     6,
           680,  6467,   391,    26,    34,   808,     5,  3508,    23,   172,
            64, 38177,   420, 22452,     2,  1705,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0],
        [ 2714,    47,   619,    11,   133,   367,     7,   913,  4967,     7,
           478,   230,    22,    47,   619,     8,  2714,    13,    10,   478,
             4,  4317,     4,    11,     3,  2087,   133,     7,    31,   366,
          4967,  4070,    25,  4569,   639,    11,     3,   151,     2,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0

/home/orr/Documentos/ULPGC/4.Cuarto/PLN/Glove/.venv/lib/python3.11/site-packages/torch/utils/data/datapipes/iter/combining.py:297: UserWarning: Some child DataPipes are not exhausted when __iter__ is called. We are resetting the buffer and each child DataPipe will read from the start again.
  warnings.warn("Some child DataPipes are not exhausted when __iter__ is called. We are resetting "


#### **Entrenamiento**

Vamos a construir las funciones <code>train</code> y <code>evaluate</code>. Dentro de la función <code>evaluate</code> vamos a prestar atención al contexto <code>with torch.no_grad():</code>, que sirve para indicar que no se van a calcular los gradientes. Esto es así porque en la fase de evaluación no se van a actualizar los pesos de la red, solo se van a utilizar para calcular la precisión de la red. Esto optimiza los recursos utilizados por PyTorch en cuanto a la gestión de la memoria.

Recuerda que en PyTorch, `model.train()` y `model.eval()` son métodos que se utilizan para cambiar el modo de entrenamiento de un modelo de aprendizaje profundo. Estos métodos afectan el comportamiento de ciertos módulos en el modelo, como las capas de dropout y normalización, que se comportan de manera diferente durante el entrenamiento y la evaluación. Específicamente hacen lo siguiente:

**model.train()**:
   - Cuando se llama a `model.train()`, el modelo se coloca en modo de entrenamiento.
   - En este modo, las capas de dropout se activan, lo que significa que se aplican durante el entrenamiento para ayudar a evitar el sobreajuste. Durante el entrenamiento, las capas de dropout "apagan" aleatoriamente un porcentaje de las unidades (neuronas) en la red en cada paso de entrenamiento.
   - También afecta a las capas de normalización, como Batch Normalization, para que utilicen estadísticas de mini-lotes (batches) durante el entrenamiento.

**model.eval()**:
   - Cuando se llama a `model.eval()`, el modelo se coloca en modo de evaluación o inferencia.
   - En este modo, las capas de dropout se desactivan, lo que significa que no se aplican durante la evaluación. Esto garantiza que el modelo produzca resultados deterministas y coherentes durante la inferencia.
   - Las capas de normalización, como Batch Normalization, utilizan estadísticas acumuladas durante el entrenamiento (en lugar de estadísticas de mini-lotes) para garantizar una evaluación coherente y precisa.


En la práctica, es común utilizar `model.train()` antes de cada época de entrenamiento y `model.eval()` antes de realizar inferencia o evaluación en un modelo entrenado. Cambiar entre estos modos es esencial para garantizar que el modelo se comporte correctamente en diferentes etapas del proceso de entrenamiento y evaluación.

In [22]:
import time

# Hyperparameters
EPOCHS = 10  # epoch
LR = 5  # learning rate

criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=LR)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1.0, gamma=0.1)

def train(dataloader):
    model.train()
    total_acc, total_count, max_acc = 0, 0, 0
    log_interval = 500
    start_time = time.time()

    for idx, (label, text) in enumerate(dataloader):
        optimizer.zero_grad()
        predicted_label = model(text)
        loss = criterion(predicted_label, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.1)
        optimizer.step()

        total_acc += (predicted_label.argmax(1) == label).sum().item()
        total_count += label.size(0)
        if idx % log_interval == 0 and idx > 0:
            elapsed = time.time() - start_time
            print('| {:5d} batches '
                  '| accuracy {:8.3f}'.format(idx, total_acc / total_count))

            if max_acc < total_acc / total_count:
                max_acc = total_acc / total_count
                
            total_acc, total_count = 0, 0
            start_time = time.time()
    return max_acc


def evaluate(dataloader):
    model.eval()
    total_acc, total_count = 0, 0

    with torch.no_grad():
        for idx, (label, text) in enumerate(dataloader):
            predicted_label = model(text)
            loss = criterion(predicted_label, label)
            total_acc += (predicted_label.argmax(1) == label).sum().item()
            total_count += label.size(0)
    return total_acc / total_count

In [ ]:
for epoch in range(1, EPOCHS + 1):
    epoch_start_time = time.time()
    accu_train = train(train_dataloader)
    accu_val = evaluate(test_dataloader)
    
    print("-" * 59)
    print(
        "| end of epoch {:3d} | time: {:5.2f}s | "
        "valid accuracy {:8.3f} ".format(
            epoch, time.time() - epoch_start_time, accu_val
        )
    )
    print("-" * 59)

|   500 batches | accuracy    0.248
|  1000 batches | accuracy    0.255
|  1500 batches | accuracy    0.307
-----------------------------------------------------------
| end of epoch   1 | time: 49.02s | valid accuracy    0.441 
-----------------------------------------------------------
|   500 batches | accuracy    0.428
|  1000 batches | accuracy    0.456
|  1500 batches | accuracy    0.464
-----------------------------------------------------------
| end of epoch   2 | time: 42.55s | valid accuracy    0.483 
-----------------------------------------------------------
|   500 batches | accuracy    0.479
|  1000 batches | accuracy    0.542
|  1500 batches | accuracy    0.639
-----------------------------------------------------------
| end of epoch   3 | time: 34.99s | valid accuracy    0.716 
-----------------------------------------------------------
|   500 batches | accuracy    0.755
|  1000 batches | accuracy    0.796
|  1500 batches | accuracy    0.822
-------------------------

---

## **Práctica 1: Uso de *embeddings* preentrenados**

Modifica el código anterior para adaptar el modelo LSTM al uso de *embeddings* preentrenados. Para ello, usa <code>from torchtext.vocab import GloVe</code> y elige el conjunto de *embeddings* GloVe que prefieras. Puedes encontrar más información en https://pytorch.org/text/stable/vocab.html#torchtext.vocab.GloVe

Verifica si se produce una mejora en la precisión del modelo. ¿Qué ocurre si usas un conjunto de *embeddings* preentrenados de diferentes tamaños?

#### ¿Qué es GloVe?

__GloVe__ (_Global Vectors for Word Representation_) es un método para generar vectores de palabras basándose en una matriz de co-ocurrencia que mide la frecuencia con la que estas aparecen juntas en un corpus. Así, consigue capturar relaciones semánticas y sintácticas de manera global. A diferencia de métodos más recientes, GloVe produce _embeddings_ estáticos, lo que implica que cada palabra tiene un único vector, independientemente del contexto. Además, estos vectores pueden tener diferentes dimensiones (p. ej., 50, 100, 200 o 300), donde las mayores capturan más matices semánticos a costa de requerir más recursos de cómputo.


In [ ]:
from torchtext.vocab import GloVe

class LSTMTextClassificationModel(nn.Module):
    def __init__(self, pretrained_embeddings, embed_dim, hidden_dim, num_class):
        super(LSTMTextClassificationModel, self).__init__()
        self.embedding = nn.Embedding.from_pretrained(pretrained_embeddings, freeze=True) # Changed to use pretrained embeddings
        self.lstm = nn.LSTM(embed_dim, hidden_dim, batch_first=True)
        self.fc = nn.Linear(hidden_dim, num_class)

    def forward(self, text):
        embedded = self.embedding(text)
        lstm_out, _ = self.lstm(embedded)
        last_output = lstm_out[:, -1, :]
        output = self.fc(last_output)
        return output

def build_embedding_matrix(vocab, glove_dim, glove_name='6B', glove_cache='./.vector_cache'):
    glove = GloVe(name=glove_name, dim=glove_dim, cache=glove_cache)

    num_embeddings = len(vocab)
    embedding_dim = glove.dim
    embedding_matrix = torch.zeros((num_embeddings, embedding_dim))

    for idx, token in enumerate(vocab.get_itos()):
        if token in glove.stoi:  
            embedding_matrix[idx] = glove[token]
        else:  
            embedding_matrix[idx] = torch.randn(embedding_dim)

    return embedding_matrix

def train(dataloader, model, criterion, optimizer, scheduler):
    model.train()
    total_acc, total_count, max_acc = 0, 0, 0
    log_interval = 500
    start_time = time.time()

    for idx, (label, text) in enumerate(dataloader):
        optimizer.zero_grad()
        predicted_label = model(text)
        loss = criterion(predicted_label, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.1)
        optimizer.step()

        total_acc += (predicted_label.argmax(1) == label).sum().item()
        total_count += label.size(0)
        if idx % log_interval == 0 and idx > 0:
            elapsed = time.time() - start_time
            print('| {:5d} batches '
                  '| accuracy {:8.3f}'.format(idx, total_acc / total_count))

            if max_acc < total_acc / total_count:
                max_acc = total_acc / total_count
                
            total_acc, total_count = 0, 0
            start_time = time.time()
    return max_acc


def evaluate(dataloader, model, criterion, predicted_label, label):
    model.eval()
    total_acc, total_count = 0, 0

    with torch.no_grad():
        for idx, (label, text) in enumerate(dataloader):
            predicted_label = model(text)
            loss = criterion(predicted_label, label)
            total_acc += (predicted_label.argmax(1) == label).sum().item()
            total_count += label.size(0)
    return total_acc / total_count

In [ ]:
def execute(glove_dim=100, epochs=5):
    embedding_matrix = build_embedding_matrix(vocab, glove_dim=glove_dim)

    model = LSTMTextClassificationModel(embedding_matrix, glove_dim, 64, 4) # Change deppending on the embedding dimension
    model.train()

    for batch in train_dataloader:
        predicted_label = model(batch[1])
        label = batch[0]
        break

    EPOCHS = epochs  # epoch
    LR = 5  # learning rate

    criterion = torch.nn.CrossEntropyLoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=LR)
    scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1.0, gamma=0.1)

    best_val_acc = 0
    for epoch in range(1, EPOCHS + 1):
        epoch_start_time = time.time()
        accu_train = train(train_dataloader, model, criterion, optimizer, scheduler)
        accu_val = evaluate(test_dataloader, model, criterion, predicted_label, label)
    
        print("-" * 59)
        print(
            "| end of epoch {:3d} | time: {:5.2f}s | "
            "valid accuracy {:8.3f} ".format(
            epoch, time.time() - epoch_start_time, accu_val
        )
        )
        print("-" * 59)
        if accu_val > best_val_acc:
            best_val_acc = accu_val

    print(f"\nMejor accuracy en validación: {best_val_acc:.3f}")

In [25]:
execute(glove_dim=50, epochs=10)

/home/orr/Documentos/ULPGC/4.Cuarto/PLN/Glove/.venv/lib/python3.11/site-packages/torch/utils/data/datapipes/iter/combining.py:297: UserWarning: Some child DataPipes are not exhausted when __iter__ is called. We are resetting the buffer and each child DataPipe will read from the start again.
  warnings.warn("Some child DataPipes are not exhausted when __iter__ is called. We are resetting "


|   500 batches | accuracy    0.253
|  1000 batches | accuracy    0.255
|  1500 batches | accuracy    0.557
-----------------------------------------------------------
| end of epoch   1 | time: 46.76s | valid accuracy    0.684 
-----------------------------------------------------------
|   500 batches | accuracy    0.690
|  1000 batches | accuracy    0.692
|  1500 batches | accuracy    0.749
-----------------------------------------------------------
| end of epoch   2 | time: 38.90s | valid accuracy    0.862 
-----------------------------------------------------------
|   500 batches | accuracy    0.868
|  1000 batches | accuracy    0.879
|  1500 batches | accuracy    0.894
-----------------------------------------------------------
| end of epoch   3 | time: 39.96s | valid accuracy    0.879 
-----------------------------------------------------------
|   500 batches | accuracy    0.883
|  1000 batches | accuracy    0.887
|  1500 batches | accuracy    0.903
-------------------------

In [26]:
execute(glove_dim=100, epochs=10)

|   500 batches | accuracy    0.257
|  1000 batches | accuracy    0.298
|  1500 batches | accuracy    0.785
-----------------------------------------------------------
| end of epoch   1 | time: 48.40s | valid accuracy    0.874 
-----------------------------------------------------------
|   500 batches | accuracy    0.873
|  1000 batches | accuracy    0.888
|  1500 batches | accuracy    0.900
-----------------------------------------------------------
| end of epoch   2 | time: 50.55s | valid accuracy    0.888 
-----------------------------------------------------------
|   500 batches | accuracy    0.891
|  1000 batches | accuracy    0.896
|  1500 batches | accuracy    0.909
-----------------------------------------------------------
| end of epoch   3 | time: 52.06s | valid accuracy    0.901 
-----------------------------------------------------------
|   500 batches | accuracy    0.897
|  1000 batches | accuracy    0.905
|  1500 batches | accuracy    0.918
-------------------------

In [27]:
execute(glove_dim=300, epochs=10)

|   500 batches | accuracy    0.455
|  1000 batches | accuracy    0.850
|  1500 batches | accuracy    0.892
-----------------------------------------------------------
| end of epoch   1 | time: 72.86s | valid accuracy    0.888 
-----------------------------------------------------------
|   500 batches | accuracy    0.889
|  1000 batches | accuracy    0.899
|  1500 batches | accuracy    0.914
-----------------------------------------------------------
| end of epoch   2 | time: 62.61s | valid accuracy    0.874 
-----------------------------------------------------------
|   500 batches | accuracy    0.904
|  1000 batches | accuracy    0.911
|  1500 batches | accuracy    0.925
-----------------------------------------------------------
| end of epoch   3 | time: 52.29s | valid accuracy    0.904 
-----------------------------------------------------------
|   500 batches | accuracy    0.914
|  1000 batches | accuracy    0.921
|  1500 batches | accuracy    0.932
-------------------------

### Resultados:

| **Embeddings** | **Precisión Inicial** | **Precisión Final** | **Mejor Precisión Validación** | **Tiempo Promedio por Epoch** |
|-----------|------------------------|----------------------|---------------------------------|--------------------------------|
| Sin GloVe (25) | 0.441                 | 0.879               | 0.879                          | 34-49 segundos                |
| GloVe 50  | 0.684                 | 0.904               | 0.904                          | 31-46 segundos                |
| GloVe 100 | 0.874                 | 0.909               | 0.909                          | 44-80 segundos                |
| GloVe 300 | 0.888                 | 0.914               | 0.918                          | 52-72 segundos                |



### **Análisis Comparativo**
   
   - A medida que aumenta la dimensión de los *embeddings*, la precisión inicial mejora significativamente, lo que demuestra que los vectores más ricos capturan mejor la semántica desde el inicio.

   - GloVe 300 logra la mayor precisión final, pero las diferencias entre GloVe 100 y GloVe 300 son pequeñas (0.5%).

   - La mejor precisión de validación (0.918) se logra con GloVe 300.

   - El costo computacional aumenta con la dimensión de los *embeddings*, haciendo que GloVe 300 sea el más costoso en tiempo de entrenamiento.